PaddleHQ/transaction-search
# Feature/tools 291 reduce required parameters

### Ticket(s):
Link to ticket:
https://paddle.atlassian.net/browse/TOOLS-291

### Checklist

### What has been done:
1. Ugly hacking about in the initial transaction queries:
    1. The non-flagged transactions are retrieved by different queries depending on which input parameters are given. We also try to speed up the query conditionally, based on whether the last four digits were given.
    1. The flagged transactions are also retrieved by different queries depending on the input parameters, but there is no need to attempt to speed it up: the transactions are always card transactions so the join to `spreedly_log` makes sense.
1. The input parameters are validated according to the new logic.

### How to test:

Note: on staging there is a lot of broken test data. Ledgers exist but not the matching orders. If this occurs, try to create an order with a more specific amount.

1. Deploy to staging.
1. Create a PayPal transaction with a reasonably unique amount.
1. Search for your transaction with a wide date range (e.g. a month) and just the amount.
1. Confirm that your transaction is included.
1. Create a card transaction with a reasonably unique amount.
1. Search for your transaction with a wide date range (e.g. a month) and just the amount.
1. Confirm that your transaction is included.
1. Search for your transaction with a narrow date range (e.g. a day) and just the last 4 digits.
1. Confirm that your transaction is included.
1. Search for your transaction with a wide date range (e.g. a day), the last 4 digits and amount.
1. Confirm that your transaction is included.
